### PR TITLE
[SofaBaseMechanics] Optimize barycentric mapping initialization

### PR DESCRIPTION
**Pull request by EulalieCoevoet**
See issue #784 : Quadractic complexity in BarycentricMapping 
This PR depends on PR #797: Cleaning and refactoring of BarycentricMapping

In this PR the initialization of BarycentricMapping (for topology container only) is optimized using spatial hashing. 
______________________________________________________

This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by damienmarchal** [comment]
+1

I will give it a look.

**Comment by damienmarchal** [comment]
On my PC this version does not compile...do the PR 797 is required to be merged before this one ?
On the CI the problem is probably caused by the Cuda version that is not refactored.

**Comment by EulalieCoevoet** [comment]
No, the PR #797 is included in this one.  
I just wanted to split both contributions to ease the review somehow.  
It compiles on my two computers, isn't it a problem with plugins? The cleaning indeed could break some not updated code.

**Comment by damienmarchal** [comment]
Yeah...after compilation issues, link issues :)

**Comment by damienmarchal** [comment]
Hi all,

I think Eulalie's work is done. 

Your feedback & review are  welcome because it is a big PR.

**Comment by epernod** [comment]
[ci-build][with-all-tests]

File: SofaKernel/modules/SofaBaseMechanics/BarycentricMapping.h [hunk 1]
```diff
@@ -599,6 +610,33 @@ class BarycentricMapperTopologyContainer : public TopologyBarycentricMapper<In,O
...
 
+    unsigned int getHashIndexFromCoord(const Vector3& x)
+    {
+        Vec3i v = getGridIndices(x);
+        return getHashIndexFromIndices(v[0],v[1],v[2]);
+    }
+
+    unsigned int getHashIndexFromIndices(const int& x, const int& y, const int& z)
+    {
+        unsigned int h = (73856093*x^19349663*y^83492791*z)%m_hashTableSize;
```
Comment (epernod) [inline SofaKernel/modules/SofaBaseMechanics/BarycentricMapping.h]: magic numbers?

File: SofaKernel/modules/SofaBaseMechanics/BarycentricMapping.inl [hunk 2]
```diff
@@ -719,9 +719,123 @@ int BarycentricMapperHexahedronSetTopology<In,Out,MappingDataType,Element>::setP
...
+    {
+        Element element = elements[i];
+        Vector3 min=in[element[0]], max=in[element[0]];
+
+        for(unsigned int j=0; j<element.size(); j++)
+        {
+            for(int k=0; k<3; k++)
+            {
+                if(in[element[j]][k]<min[k]) min[k]=in[element[j]][k];
+                if(in[element[j]][k]>max[k]) max[k]=in[element[j]][k];
```
Comment (epernod) [inline SofaKernel/modules/SofaBaseMechanics/BarycentricMapping.inl]: speaking about optimisation, small detail, it is better to define :
int elemid = element[j]; 
before the loop on k.

**Comment by epernod** [comment]
Hi again @EulalieCoevoet 
I tried to check the difference with the PR #797
I suppose it is this commit: https://github.com/sofa-framework/sofa/pull/798/commits/5f56d38ac531e7dfd8a0d63eb7d00ddc870c21ea
Could you add in 2 lines in the header, what criteria is used to create the hash table and then which method is used to find element in the table. If already done I apologize, the changes log is very hard to read.

**Review by epernod — COMMENTED** [review]
https://github.com/sofa-framework/sofa/pull/798/commits/5f56d38ac531e7dfd8a0d63eb7d00ddc870c21ea

File: SofaKernel/modules/SofaBaseMechanics/BarycentricMapping.h (hunk 1 above)
Comment (damienmarchal) [reply SofaKernel/modules/SofaBaseMechanics/BarycentricMapping.h]: if prime numbers are magic then yes :)
https://stackoverflow.com/questions/5928725/hashing-2d-3d-and-nd-vectors

Comment (epernod) [reply SofaKernel/modules/SofaBaseMechanics/BarycentricMapping.h]: so big prime number. I take it for magic !

**Review by epernod — APPROVED** [review]

**Pull request merged**